File: dashboards/django/views/skills.py
"""
Skills view: browse builtin skills and manage installed community skills.
"""

from django.http import HttpResponse, HttpResponseRedirect
from django.template.loader import render_to_string

from utils.api import api_request
from utils.helpers import badge, page_context


def skills_view(request):
    """Handle skill browsing, enabling, disabling, and uninstalling."""
    if not request.session.get('token'):
        return HttpResponseRedirect('/login')

    token = request.session['token']

    if request.method == 'POST':
        action = request.POST.get('_action')
        skill_id = request.POST.get('skill_id')
        if action == 'enable':
            result = api_request('POST', f'/engine/community/{skill_id}/enable', token=token)
            request.session['flash'] = 'Skill enabled!' if not result.get('error') else result.get('error', 'Failed')
        elif action == 'disable':
            result = api_request('POST', f'/engine/community/{skill_id}/disable', token=token)
            request.session['flash'] = 'Skill disabled!' if not result.get('error') else result.get('error', 'Failed')
        elif action == 'uninstall':
            result = api_request('DELETE', f'/engine/community/{skill_id}', token=token)
            request.session['flash'] = 'Skill uninstalled!' if not result.get('error') else result.get('error', 'Failed')
        return HttpResponseRedirect('/skills')

    categories_data = api_request('GET', '/engine/skills/by-category', token=token)
    installed_data = api_request('GET', '/engine/community/installed?orgId=default', token=token)

    categories = categories_data.get('categories', {})
    installed = installed_data.get('skills', [])

    for s in installed:
        if s.get('enabled'):
            s['status_badge'] = badge('enabled', 'b-a')
        else:
            s['status_badge'] = badge('disabled', 'b-r')

    ctx = page_context(request, 'Skills', 'skills')
    ctx['categories'] = categories
    ctx['installed'] = installed

    html = render_to_string('skills.html', ctx)
    return HttpResponse(html)

File: dashboards/rails/routes/skills.rb
# frozen_string_literal: true

# Skills Controller

get '/skills' do
  require_auth!
  categories_res = api_get('/engine/skills/by-category')
  @categories = categories_res[:status] == 200 ? (categories_res[:body]['categories'] || categories_res[:body]) : {}
  @categories = @categories.is_a?(Hash) ? @categories : {}
  installed_res = api_get('/engine/community/installed?orgId=default')
  @installed = installed_res[:status] == 200 ? (installed_res[:body]['skills'] || installed_res[:body]) : []
  @installed = @installed.is_a?(Array) ? @installed : []
  erb :skills
end

post '/skills/enable' do
  require_auth!
  skill_id = params[:skill_id]
  res = api_post("/engine/community/#{skill_id}/enable", {})
  set_flash(res[:status] < 300 ? 'Skill enabled' : (res[:body].is_a?(Hash) ? res[:body]['error'] || 'Failed' : 'Failed'), res[:status] < 300 ? 'success' : 'danger')
  redirect '/skills'
end

post '/skills/disable' do
  require_auth!
  skill_id = params[:skill_id]
  res = api_post("/engine/community/#{skill_id}/disable", {})
  set_flash(res[:status] < 300 ? 'Skill disabled' : (res[:body].is_a?(Hash) ? res[:body]['error'] || 'Failed' : 'Failed'), res[:status] < 300 ? 'success' : 'danger')
  redirect '/skills'
end

post '/skills/uninstall' do
  require_auth!
  skill_id = params[:skill_id]
  res = api_delete("/engine/community/#{skill_id}")
  set_flash(res[:status] < 300 ? 'Skill uninstalled' : (res[:body].is_a?(Hash) ? res[:body]['error'] || 'Failed' : 'Failed'), res[:status] < 300 ? 'success' : 'danger')
  redirect '/skills'
end
